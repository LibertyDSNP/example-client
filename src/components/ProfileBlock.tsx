import React from "react";
import LoginSetupInstructions from "./LoginSetupInstructions";
import Profile from "./Profile";
import { useAppSelector } from "../redux/hooks";

const ProfileBlock = (): JSX.Element => {
  const userId: string | undefined = useAppSelector((state) => state.user.id);

  const displayId: string | undefined = useAppSelector(
    (state) => state.user.displayId
  );

  const displayProfileBlockElement = () => {
    if (userId) {
      return <Profile />;
    }
    return displayId ? <Profile /> : <LoginSetupInstructions />;
  };

  return (
    <div className="ProfileBlock__block">{displayProfileBlockElement()}</div>
  );
};
export default ProfileBlock;
