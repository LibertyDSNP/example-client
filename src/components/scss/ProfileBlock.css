/* Error: Undefined variable.
 *   ,
 * 2 |   background-color: $white;
 *   |                     ^^^^^^
 *   '
 *   ProfileBlock.scss 2:21  root stylesheet */

body::before {
  font-family: "Source Code Pro", "SF Mono", Monaco, Inconsolata, "Fira Mono",
      "Droid Sans Mono", monospace, monospace;
  white-space: pre;
  display: block;
  padding: 1em;
  margin-bottom: 1em;
  border-bottom: 2px solid black;
  content: "Error: Undefined variable.\a   \2577 \a 2 \2502    background-color: $white;\a   \2502                      ^^^^^^\a   \2575 \a   ProfileBlock.scss 2:21  root stylesheet";
}
