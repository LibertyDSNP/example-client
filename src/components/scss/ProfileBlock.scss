.ProfileBlock__block {
  background-color: $white;
  padding: 32px;
  display: flex;
  flex-direction: column;
  flex-basis: 30%;
  max-width: 475px;
  height: min-content;
  border-radius: 25px;
}

.ProfileBlock__personalInfoBlock {
  display: flex;
  justify-content: space-between;
}

.ProfileBlock__avatarBlock {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.ProfileBlock__personalInfo {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  margin: 0 0 12px 16px;
  height: 172px;
}

.ProfileBlock__personalInfoLabel {
  color: $text-gray;
  line-height: 1;
}

.ProfileBlock__name {
  @include large-text;
  @include bold-text;
  background: none;
  border-color: $clear;
  line-height: 1;
  margin-bottom: 16px;
}

.ProfileBlock__handle {
  @include medium-text;
  @include bold-text;
  background: none;
  border-color: $clear;
  line-height: 1;
  margin-bottom: 16px;
}

.ProfileBlock__socialAddress {
  @include small-text;
  @include bold-text;
  background: none;
  border-color: $clear;
  line-height: 1;
}

.ProfileBlock__editButton {
  width: 70px;
  margin-top: 12px;
  text-transform: uppercase;
}

.ProfileBlock__editing {
  background-color: $gray;

  &:focus {
    outline: none !important;
  }
}

.ProfileBlock__loginText {
  @include large-text;
  color: $dark-gray;
  line-height: 1.3;
}
