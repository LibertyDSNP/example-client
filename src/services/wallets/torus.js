import web3Torus from "./tweb3";

export const enableTorus = async (buildEnv) => {
  await web3Torus.initialize(buildEnv || "Testing");
};

export const isInitialized = () => {
  return web3Torus.initialized;
};

export const getUserInfo = async () => {
  const userInfo = await web3Torus.torus.getUserInfo();
  return userInfo;
};

export const logout = () => {
  web3Torus.torus.cleanUp().then(() => {
    sessionStorage.setItem("pageUsingTorus", false);
  });
};

export const getPublicAddress = async (verifier, verifierId) => {
  return await web3Torus.torus.getPublicAddress({
    verifier,
    verifierId,
    isExtended: true,
  });
};

export const getWalletAddress = async () => {
  const { verifier, verifierId } = await getUserInfo();
  return (await getPublicAddress(verifier, verifierId)).address;
};

export const getAccount = async () => {
  return await web3Torus.torus.accounts;
};

export const getBalance = (account) => {
  return web3Torus.web3.eth.getBalance(account).then((balance) => {
    return balance;
  });
};
