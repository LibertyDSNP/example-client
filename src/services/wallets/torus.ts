import { HexString } from "../../utilities/types";
import web3Torus, { BuildEnvironment } from "./tweb3";

export const enableTorus = async (
  buildEnv?: BuildEnvironment
): Promise<void> => {
  await web3Torus.initialize(buildEnv || "testing");
};

export const isInitialized = (): boolean => {
  return web3Torus.initialized;
};

interface UserInfo {
  email: string;
  name: string;
  profileImage: string;
  verifier: string;
  verifierId: string;
}

export const getUserInfo = async (): Promise<UserInfo> => {
  const userInfo = await web3Torus.torus?.getUserInfo(
    "This site is requesting your information"
  );
  if (!userInfo) throw new Error("Unable to find user's info");
  return userInfo;
};

export const logout = (): void => {
  web3Torus.torus?.cleanUp().then(() => {
    sessionStorage.setItem("pageUsingTorus", "false");
  });
};

type VerifierTypes = "google" | "reddit" | "discord";
type AddressType = undefined | string | { address: string };
export const getPublicAddress = async (
  verifier: VerifierTypes,
  verifierId: string
): Promise<AddressType> => {
  return await web3Torus.torus?.getPublicAddress({
    verifier,
    verifierId,
    isExtended: true,
  });
};

export const getWalletAddress = async (): Promise<string> => {
  const { verifier, verifierId } = await getUserInfo();
  const walletObject = await getPublicAddress(
    verifier as VerifierTypes,
    verifierId
  );
  if (!walletObject) throw new Error("No Wallet Address found for User");
  else if (typeof walletObject === "string") return walletObject;
  else return walletObject.address;
};

export const getBalance = async (walletAddress: HexString): Promise<string> => {
  return await web3Torus.web3.eth.getBalance(walletAddress).then((balance) => {
    return balance;
  });
};
