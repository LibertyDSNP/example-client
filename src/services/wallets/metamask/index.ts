import * as metamask from "./metamask";
import { Wallet } from "../wallet";

const metamaskWallet: Wallet = {
  login: async () => {
    return metamask.getWalletAddress();
  },
  logout: () => {
    return;
  },
  getAddress: async () => {
    return metamask.getWalletAddress();
  },
  getWeb3: () => {
    return metamask.getWeb3();
  },
};

export default metamaskWallet;
