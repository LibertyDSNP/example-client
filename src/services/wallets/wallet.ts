import { HexString } from "../../utilities/types";

import * as torus from "./torus/torus";
import * as metamask from "./metamask/metamask";
import Web3 from "web3";

export enum WalletType {
  TORUS = "TORUS",
  METAMASK = "METAMASK",
  // Add new WalletTypes Here
}

interface Wallet {
  login: () => Promise<HexString>;
  logout: () => void;
  getAddress: () => Promise<HexString>;
  getBalance: (walletAddress: HexString) => Promise<number>;
  getWeb3: () => Web3;
}

export const wallet = (walletType: WalletType): Wallet => {
  switch (walletType) {
    case WalletType.TORUS:
      return torusWallet;
    case WalletType.METAMASK:
      return metamaskWallet;
    // Add new WalletTypes Here
  }
};

const torusWallet: Wallet = {
  login: async () => {
    await torus.enableTorus();
    return await torus.getWalletAddress();
  },
  logout: () => {
    if (torus.isInitialized()) {
      torus.logout();
    }
  },
  getAddress: async () => {
    return await torus.getWalletAddress();
  },
  getBalance: async (walletAddress: HexString) => {
    const stringBal = await torus.getBalance(walletAddress);
    return Number(stringBal);
  },
  getWeb3: () => {
    return torus.getWeb3();
  },
};

const metamaskWallet: Wallet = {
  login: async () => {
    return metamask.getWalletAddress();
  },
  logout: () => {
    return;
  },
  getAddress: async () => {
    return metamask.getWalletAddress();
  },
  getBalance: async (walletAddress: HexString) => {
    const stringBal = await metamask.getBalance(walletAddress);
    return Number(stringBal);
  },
  getWeb3: () => {
    return metamask.getWeb3();
  },
};

// Add new Wallet interface here
