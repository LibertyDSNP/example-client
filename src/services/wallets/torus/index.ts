import * as torus from "./torus";
import { Wallet } from "../wallet";

const torusWallet: Wallet = {
  login: async () => {
    await torus.enableTorus();
    return await torus.getWalletAddress();
  },
  logout: () => {
    if (torus.isInitialized()) {
      torus.logout();
    }
  },
  getAddress: async () => {
    return await torus.getWalletAddress();
  },
  getWeb3: () => {
    return torus.getWeb3();
  },
};

export default torusWallet;
