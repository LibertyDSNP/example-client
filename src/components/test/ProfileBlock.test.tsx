import ProfileBlock from "../ProfileBlock";
import { mount } from "enzyme";
import { getPrefabProfile } from "../../test/testProfiles";
import { componentWithStore, createMockStore } from "../../test/testhelpers";
import { getPreFabSocialGraph } from "../../test/testGraphs";
import * as wallet from "../../services/wallets/wallet";
import { waitFor } from "@testing-library/react";

const id = "0x03f2";
const displayedProfileId = id;
const otherId = "0x1234";
const profile = getPrefabProfile(0);
const otherProfile = getPrefabProfile(1);
const graphs = getPreFabSocialGraph();
const walletType = wallet.WalletType.TORUS;
const store = createMockStore({
  user: { id, walletType, displayedProfileId },
  profiles: { profiles: { [id]: profile, [otherId]: otherProfile } },
  graphs: graphs,
});

describe("Profile Block", () => {
  it("renders without crashing", async () => {
    await waitFor(() => {
      expect(() => {
        mount(componentWithStore(ProfileBlock, store));
      }).not.toThrow();
    });
  });

  it("displays login prompt when not logged in", () => {
    const store = createMockStore({
      user: { undefined },
      profiles: { profiles: {} },
      graphs: { graphs: graphs },
      displayedProfileId: undefined,
    });
    const component = mount(componentWithStore(ProfileBlock, store));
    expect(component.find("ProfileBlock").text()).toContain(
      "Login Quick Start"
    );
  });

  it("editable on edit button click", async () => {
    const component = mount(componentWithStore(ProfileBlock, store));
    component.find(".ProfileBlock__editButton").first().simulate("click");
    await waitFor(() => {
      expect(component.find(".ProfileBlock__name").props().disabled).toEqual(
        false
      );
    });
  });

  describe("save button click", () => {
    it("save button disabled when not edited", async () => {
      const component = mount(componentWithStore(ProfileBlock, store));
      component.find(".ProfileBlock__editButton").first().simulate("click");
      await waitFor(() => {
        expect(
          component.find(".ProfileBlock__editButton").first().props().disabled
        ).toEqual(true);
      });
    });
    it("save button enabled when edited", async () => {
      const component = mount(componentWithStore(ProfileBlock, store));
      component.find(".ProfileBlock__editButton").first().simulate("click");
      component
        .find(".ProfileBlock__name")
        .simulate("change", { target: { value: "Monday NewLastName" } });
      await waitFor(() => {
        expect(
          component.find(".ProfileBlock__editButton").first().props().disabled
        ).toEqual(false);
      });
    });
  });

  it("cancels on cancel button click", async () => {
    const component = mount(componentWithStore(ProfileBlock, store));
    component.find(".ProfileBlock__editButton").first().simulate("click");
    component
      .find(".ProfileBlock__name")
      .simulate("change", { target: { value: "Monday TestLastName" } });
    await waitFor(() => {
      expect(component.find(".ProfileBlock__name").props().value).toEqual(
        "Monday TestLastName"
      );
    });
    component.find(".ProfileBlock__editButton").last().simulate("click");
    expect(component.find(".ProfileBlock__name").props().value).toEqual(
      "Monday January"
    );
  });

  describe("display other user", () => {
    const otherUserStore = createMockStore({
      user: { id, walletType, otherId },
      profiles: { profiles: { [id]: profile, [otherId]: otherProfile } },
      graphs: graphs,
      displayedProfileId: { otherId },
    });

    it("does not show edit button if other user", async () => {
      const component = mount(componentWithStore(ProfileBlock, otherUserStore));
      expect(component.find(".ProfileBlock__editButton")).toEqual({});
    });
  });
});
