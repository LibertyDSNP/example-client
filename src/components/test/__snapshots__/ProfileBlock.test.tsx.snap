// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Block display other user shows other user's profile 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <ProfileBlock>
        <div
          className="ProfileBlock__block"
        >
          <Profile>
            <div
              className="ProfileBlock__personalInfoBlock"
            >
              <div
                className="ProfileBlock__avatarBlock"
              >
                <UserAvatar
                  avatarSize="large"
                >
                  <Avatar
                    alt="anonymous"
                    className="UserAvatar"
                    shape="circle"
                    size={100}
                  >
                    <span
                      className="ant-avatar ant-avatar-circle UserAvatar"
                      style={
                        Object {
                          "fontSize": 18,
                          "height": 100,
                          "lineHeight": "100px",
                          "width": 100,
                        }
                      }
                    >
                      <ResizeObserver
                        onResize={[Function]}
                      >
                        <span
                          className="ant-avatar-string"
                          style={
                            Object {
                              "WebkitTransform": "scale(1) translateX(-50%)",
                              "lineHeight": "100px",
                              "msTransform": "scale(1) translateX(-50%)",
                              "transform": "scale(1) translateX(-50%)",
                            }
                          }
                        />
                      </ResizeObserver>
                    </span>
                  </Avatar>
                </UserAvatar>
              </div>
              <div
                className="ProfileBlock__personalInfo"
              >
                <label
                  className="ProfileBlock__personalInfoLabel"
                >
                  NAME
                </label>
                <input
                  className="ProfileBlock__name"
                  disabled={true}
                  onChange={[Function]}
                  value="Anonymous"
                />
                <label
                  className="ProfileBlock__personalInfoLabel"
                >
                  HANDLE
                </label>
                <input
                  className="ProfileBlock__handle"
                  disabled={true}
                  onChange={[Function]}
                />
                <label
                  className="ProfileBlock__personalInfoLabel"
                >
                  SOCIAL ADDRESS
                </label>
                <input
                  className="ProfileBlock__dsnpUserId"
                  disabled={true}
                  value="Anonymous"
                />
              </div>
            </div>
            <ConnectionsList>
              <div
                className="ConnectionsList__block"
              >
                <div
                  className="ConnectionsList__buttonBlock"
                >
                  <Button
                    className="ConnectionsList__button"
                    onClick={[Function]}
                  >
                    <Wave>
                      <button
                        className="ant-btn ConnectionsList__button"
                        onClick={[Function]}
                        type="button"
                      >
                        <LoadingIcon
                          existIcon={false}
                          loading={false}
                          prefixCls="ant-btn"
                        >
                          <CSSMotion
                            motionName="ant-btn-loading-icon-motion"
                            onAppearActive={[Function]}
                            onAppearStart={[Function]}
                            onEnterActive={[Function]}
                            onEnterStart={[Function]}
                            onLeaveActive={[Function]}
                            onLeaveStart={[Function]}
                            removeOnLeave={true}
                            visible={false}
                          >
                            <DomWrapper />
                          </CSSMotion>
                        </LoadingIcon>
                        <div
                          className="ConnectionsList__buttonCount"
                          key=".0"
                        >
                          0
                        </div>
                        <span
                          key=".1"
                        >
                          Followers
                        </span>
                      </button>
                    </Wave>
                  </Button>
                  <div
                    className="ConnectionsList__buttonSeparator"
                  >
                     
                  </div>
                  <Button
                    className="ConnectionsList__button"
                    onClick={[Function]}
                  >
                    <Wave>
                      <button
                        className="ant-btn ConnectionsList__button"
                        onClick={[Function]}
                        type="button"
                      >
                        <LoadingIcon
                          existIcon={false}
                          loading={false}
                          prefixCls="ant-btn"
                        >
                          <CSSMotion
                            motionName="ant-btn-loading-icon-motion"
                            onAppearActive={[Function]}
                            onAppearStart={[Function]}
                            onEnterActive={[Function]}
                            onEnterStart={[Function]}
                            onLeaveActive={[Function]}
                            onLeaveStart={[Function]}
                            removeOnLeave={true}
                            visible={false}
                          >
                            <DomWrapper />
                          </CSSMotion>
                        </LoadingIcon>
                        <div
                          className="ConnectionsList__buttonCount"
                          key=".0"
                        >
                          0
                        </div>
                        <span
                          key=".1"
                        >
                          Following
                        </span>
                      </button>
                    </Wave>
                  </Button>
                </div>
                <ConnectionsListProfiles
                  followers={Object {}}
                  following={Object {}}
                  listStatus={0}
                />
              </div>
            </ConnectionsList>
          </Profile>
        </div>
      </ProfileBlock>
    </Router>
  </BrowserRouter>
</Provider>
`;
